### PR TITLE
Add evpn ethernet-segment support for LAG interface

### PR DESCRIPTION
**Pull request by mingjunzhang2019**
##### SUMMARY
This is to add evpn ethernet-segment support for LAG interface.

##### GitHub Issues
GitHub Issue #379 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
sonic_lag_interfaces

##### OUTPUT
[regression.log](https://github.com/user-attachments/files/16531877/regression.log)

[regression-2024-08-07-11-45-15.html.pdf](https://github.com/user-attachments/files/16531880/regression-2024-08-07-11-45-15.html.pdf)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Regression testing and lots of unit testing.

### CODE REVIEW

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 1]
```diff
@@ -445,6 +575,69 @@ def get_delete_lag_interfaces_requests(self, commands):
...
+                have_po = next((po for po in have if po['name'] == po_name), {})
+                have_es = have_po.get('ethernet_segment', {})
+                if not have_es:
+                    continue
+
+                del_es = False
+                del_df_pref = False
+                if delete_es:
+                    del_es = True
+                else:
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:617]: This block should interpret a request to delete all three options as a request to delete the ethernet segment. It would also make sense for a deletion request that only specifies the esi_type to be interpreted as a request to delete the the ethernet segment.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
The basic functionality and current test results look good.

The comments I am posting address some corner cases, additional or modified functionality (for explicit deletion of an entire ethernet_segment), optimization on processing time (when merging multiple ethernet_segment additions), simplification of algorithms, code reduction, and test case enhancements/modifications.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 2]
```diff
@@ -170,6 +171,11 @@ def set_state(self, want, have):
         :returns: the commands necessary to migrate the current configuration
                   to the desired configuration
         """
+        state = self._module.params['state']
+
+        self.validate_want(want, state)
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:176]: If a playbook specifies esi: MANUAL (for example), it should be error.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 3]
```diff
@@ -178,16 +184,13 @@ def set_state(self, want, have):
...
-        if state in ('overridden', 'merged', 'replaced') and not want:
-            self._module.fail_json(msg='value of config parameter must not be empty for state {0}'.format(state))
-
         if state == 'overridden':
             commands, requests = self._state_overridden(want, have, diff_members, diff_portchannels)
         elif state == 'deleted':
             commands, requests = self._state_deleted(want, have, diff)
         elif state == 'merged':
-            commands, requests = self._state_merged(want, have, diff_members, diff_portchannels)
+            commands, requests = self._state_merged(want, have, diff,
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Corrected. Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 4]
```diff
@@ -214,6 +217,21 @@ def _state_replaced(self, want, have, diff_members, diff_portchannels):
         requests = self.get_delete_lag_interfaces_requests(replaced_list)
         if requests:
             commands.extend(update_states(replaced_list, "deleted"))
+
+        es_commands, es_requests = self.get_delete_ethernet_segment_requests(replaced_list,
+                                                                             have, False)
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Changed. Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 5]
```diff
@@ -260,14 +285,16 @@ def _state_overridden(self, want, have, diff_members, diff_portchannels):
 
         return commands, requests
 
-    def _state_merged(self, want, have, diff_members, diff_portchannels):
+    def _state_merged(self, want, have, diff, diff_members, diff_portchannels):
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Removed.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 6]
```diff
@@ -347,7 +391,26 @@ def template_for_lag_deletion(self, have, delete_members, delete_portchannels, s
...
+                    po_name = cmd['name']
+                    cmd_in_cmds = next((po for po in commands if po['name'] == po_name), {})
+                    if not cmd_in_cmds:
+                        es_cmds.append(cmd)
+                if es_cmds:
+                    commands.extend(update_states(es_cmds, state_name))
+                requests.extend(es_requests)
+
         if delete_portchannels:
+            na, es_requests = self.get_delete_po_ethernet_segment_requests(delete_portchannels,
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Changed.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 7]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+        es_path = 'data/openconfig-network-instance:network-instances/network-instance=default/evpn/ethernet-segments'
+        df_pref_path = es_path + '/ethernet-segment=%s/df-election/config/preference'
+
+        for cmd in diff_members:
+            po_name = cmd['name']
+            cmd_es = cmd.get('ethernet_segment', None)
+            if cmd_es:
+                es = dict()
+                have_po = next((po for po in have if po['name'] == po_name), {})
+                have_es = have_po.get('ethernet_segment', {})
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:449]: have_po will not be None due to line 449 and 450 (return {} if not exist)

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 8]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+
+                if cmd_es.get('esi_type'):
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:451]: It is needed due to "diff" operation removes it from cmd.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 9]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+                    esi_type = cmd_es['esi_type']
+                else:
+                    esi_type = have_es['esi_type']
+
+                esi = cmd_es.get('esi')
+                if esi_type == 'auto_lacp':
+                    if esi and esi != 'AUTO':
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Done. Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 10]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+                        self._module.fail_json(msg='{0}: value of esi must be "AUTO" for esi_type {1}'.format(po_name, esi_type))
+                    esi_t = 'TYPE_1_LACP_BASED'
+                    esi = 'AUTO'
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:459]: Removed "fail_json". Kept esi = 'AUTO'. It is needed due to "diff" operation removes it from cmd.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 11]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+                elif esi_type == 'auto_system_mac':
+                    if esi and esi != 'AUTO':
+                        self._module.fail_json(msg='{0}: value of esi must be "AUTO" for esi_type {1}'.format(po_name, esi_type))
+                    esi_t = 'TYPE_3_MAC_BASED'
+                    esi = 'AUTO'
+                elif esi_type == 'ethernet_segement_id':
+                    esi_t = 'TYPE_0_OPERATOR_CONFIGURED'
+                    if not esi:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:465]: It is needed due to "diff" operation removes it from cmd.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 12]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+                        esi = have_es['esi']
+                    esi = ''.join(esi.split(':'))
+
+                if cmd_es.get('df_preference'):
+                    df_pref = cmd_es['df_preference']
+                else:
+                    df_pref = have_es['df_preference']
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:472]: It is needed due to "diff" operation removes it from cmd.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 13]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
...
+                es_requests.append(es_request)
+
+                ed_df_pref_path = df_pref_path % quote(po_name, safe='')
+                df_pref_payload = {'openconfig-network-instance:preference': df_pref}
+                df_pref_request = {'path': ed_df_pref_path, 'method': PATCH, 'data': df_pref_payload}
+                es_requests.append(df_pref_request)
+
+                es_commands.append(cmd)
+
+        return es_commands, es_requests
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:497]: esi and df_preference have different config path. It does not work to combine them.  Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 1 above)
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:612]: Yes!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 14]
```diff
@@ -445,6 +575,69 @@ def get_delete_lag_interfaces_requests(self, commands):
...
+                if not have_es:
+                    continue
+
+                del_es = False
+                del_df_pref = False
+                if delete_es:
+                    del_es = True
+                else:
+                    if cmd_es.get('esi_type'):
+                        if cmd_es.get('esi'):
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:614]: It is "diff" operation produces those result!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 15]
```diff
@@ -445,6 +575,69 @@ def get_delete_lag_interfaces_requests(self, commands):
...
+                            if have_es.get('df_preference') != DEFAULT_ES_DF_PREF:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:615]: Deleting df_preference makes  SONIC sets it to be DEFAULT VALUE.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 16]
```diff
@@ -445,6 +575,69 @@ def get_delete_lag_interfaces_requests(self, commands):
...
+                                del_df_pref = True
+                        else:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:617]: It is "diff" operation producing those result. Let's discuss it.

File: plugins/modules/sonic_lag_interfaces.py [hunk 17]
```diff
@@ -156,6 +194,9 @@
         members:
           interfaces:
             - member: Eth1/7
+        ethernet_segment:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py:201]: Changed. Thanks!

File: plugins/modules/sonic_lag_interfaces.py [hunk 18]
```diff
@@ -179,6 +220,12 @@
 #   speed 100000
 #   no shutdown
 #
+# interface PortChannel0
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py:227]: Changed. Thanks!

File: plugins/modules/sonic_lag_interfaces.py [hunk 19]
```diff
@@ -209,6 +256,9 @@
         members:
           interfaces:
             - member: Eth1/6
+        ethernet_segment:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py:263]: Changed Thanks!

File: plugins/modules/sonic_lag_interfaces.py [hunk 20]
```diff
@@ -265,6 +321,10 @@
 # Before state:
 # -------------
 # interface PortChannel 10
+#  no shutdown
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py:328]: Added. Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:176]: Okay. After considering the point you have mentioned here and discussing this in more detail, I understand the reason for this ordering.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:449]: The lines 449 and 450 above are my suggested change. The current code in this file will just hit an exception in line 449 if have_po is None at the end of line 448.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:451]: I can see why this is needed for correct handling of 'overridden' and 'replaced' state because the previously configured 'esi_type' value may have been deleted.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:459]: This sounds like the right handling, give the fact that, for 'overridden' and 'replaced' state, the previously configured 'esi' value may have been deleted.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:465]: I agree that these instructions are needed due to the fact that for 'overridden' and 'replaced' state, the previously configured 'esi' value may have been deleted.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:472]: I agree that these instructions are needed due to the fact that for 'overridden' and 'replaced' state, the previously configured 'df_preference' value may be queued for deletion.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:497]: It does work to combine them. I have verified this with Swagger.

File: plugins/modules/sonic_lag_interfaces.py [hunk 21]
```diff
@@ -74,6 +75,31 @@
...
+        suboptions:
+          esi_type:
+            description:
+              - Specifies type of Ethernet Segment Identifier.
+            required: True
+            type: str
+            choices:
+              - auto_lacp
+              - auto_system_mac
+              - ethernet_segement_id
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_lag_interfaces.py]: Fixed.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:497]: image

[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Done.

File: plugins/module_utils/network/sonic/facts/lag_interfaces/lag_interfaces.py [hunk 22]
```diff
@@ -86,20 +98,47 @@ def populate_facts(self, connection, ansible_facts, data=None):
...
                     self.merge_portchannels(objs, obj)
+
+        es_data = self.get_ethernet_segments(data)
+        for es in es_data:
+            po_name = es['ifname']
+            esi_t = es['esi_type']
+            esi = es['esi']
+            if 'df_pref' in es:
+                df_pref = es['df_pref']
+            else:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/facts/lag_interfaces/lag_interfaces.py:116]: Done

File: plugins/modules/sonic_lag_interfaces.py [hunk 23]
```diff
@@ -74,6 +75,31 @@
...
+          esi:
+            description:
+              - Specifies value of Ethernet Segment Identifier.
+                Only "AUTO" is supported for auto_lacp and auto_system_mac.
+            type: str
+          df_preference:
+            description:
+              - The preference for Designated Forwarder election method.
+            type: int
+            default: 32767
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py:106]: Removed.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply tests/regression/roles/sonic_lag_interfaces/defaults/main.yml:1]: Done.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 24]
```diff
@@ -347,7 +390,26 @@ def template_for_lag_deletion(self, have, delete_members, delete_portchannels, s
                 requests = deleteall_requests
             if requests:
                 commands.extend(update_states(delete_members, state_name))
+
+            es_commands, es_requests = self.get_delete_ethernet_segment_requests(delete_members,
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:392]: Done. Thanks.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:446]: Changed.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 25]
```diff
@@ -445,6 +575,69 @@ def get_delete_lag_interfaces_requests(self, commands):
...
+                have_es = have_po.get('ethernet_segment', {})
+                if not have_es:
+                    continue
+
+                del_es = False
+                del_df_pref = False
+                if delete_es:
+                    del_es = True
+                else:
+                    if cmd_es.get('esi_type'):
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:617]: Done. Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:618]: Done Thanks!

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:446]: It looks like it's still unchanged. Here are the current lines 435-446:

```
    def create_ethernet_segment_requests(self, diff_members, have):
        es_commands = []
        es_path = 'data/openconfig-network-instance:network-instances/network-instance=default/evpn/ethernet-segments'
        es_payload_list = list()

        for cmd in diff_members:
            po_name = cmd['name']
            cmd_es = cmd.get('ethernet_segment', None)
            if cmd_es:
                es = dict()
                have_po = next((po for po in have if po['name'] == po_name), {})
                have_es = have_po.get('ethernet_segment', {})
```

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 26]
```diff
@@ -372,6 +435,72 @@ def create_lag_interfaces_requests(self, commands):
                 requests.append(request)
         return requests
 
+    def create_ethernet_segment_requests(self, diff_members, have):
+        es_commands = []
+        es_requests = []
+        es_path = 'data/openconfig-network-instance:network-instances/network-instance=default/evpn/ethernet-segments'
+        df_pref_path = es_path + '/ethernet-segment=%s/df-election/config/preference'
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: The equivalent change has been added.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Verified.

The revised code section looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 27]
```diff
@@ -309,17 +347,23 @@ def template_for_lag_creation(self, have, diff_members, diff_portchannels, state
...
-            diff_members_remove_none = [x for x in diff_members if x["members"]]
+            diff_members_remove_none = [x for x in diff_members if x.get("members")]
             if diff_members_remove_none:
                 request = self.create_lag_interfaces_requests(diff_members_remove_none)
                 if request:
                     requests.extend(request)
                 else:
                     requests = request
+
+            es_commands, es_requests = self.create_ethernet_segment_requests(diff_members, have)
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 28]
```diff
@@ -309,17 +347,23 @@ def template_for_lag_creation(self, have, diff_members, diff_portchannels, state
...
+            if es_requests:
+                requests.extend(es_requests)
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 29]
```diff
@@ -61,6 +61,7 @@
     ERR_MSG = to_native(e)
     LIB_IMP_ERR = traceback.format_exc()
 
+DEFAULT_ES_DF_PREF = 32767
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/facts/lag_interfaces/lag_interfaces.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/lag_interfaces/lag_interfaces.py:116]: The changed code looks good.

Resolved.

File: plugins/modules/sonic_lag_interfaces.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_lag_interfaces.py]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:446]: I see now that the default value for the 'next' invocation is an empty dictionary. This makes the following line safe "as is", so the change I suggested here is not needed.

Resolved.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml [hunk 30]
```diff
@@ -86,10 +98,16 @@ tests:
           members:
             interfaces:
               - member: "{{ interface5 }}"
+          ethernet_segment:
+            esi_type: auto_system_mac
+            df_preference: 3333
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply tests/regression/roles/sonic_lag_interfaces/defaults/main.yml]: The changed code looks good.

Resolved.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply tests/regression/roles/sonic_lag_interfaces/defaults/main.yml:1]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 1 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:616]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 31]
```diff
@@ -287,15 +313,26 @@ def _state_deleted(self, want, have, diff):
...
         else:  # delete specific lag interfaces and specific portchannels
-            commands = get_diff(want, diff, TEST_KEYS)
-            commands = remove_empties_from_list(commands)
-            want_members, want_portchannels = self.diff_list_for_member_creation(commands)
-            commands, requests = self.template_for_lag_deletion(have, want_members, want_portchannels, "deleted")
+            po_commands = get_diff(want, diff, TEST_KEYS)
+            po_commands = remove_empties_from_list(po_commands)
+            want_members, want_portchannels = self.diff_list_for_member_creation(po_commands)
+
+            del_commands, del_requests = self.template_for_lag_deletion(have, want_members,
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: The changed code looks good.

Resolved.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:392]: The changed code looks good.

Resolved.

**Review by kerry-meyer — COMMENTED** [review]
I think all issues are resolved now, but I have at least one comment in a "pending" state, so I'm posting my current "pending" comments to make sure there are no remaining issues.

I expect to be approving/merging this PR after doing this final check.

**Review by kerry-meyer — APPROVED** [review]
All issues are now resolved.

The current proposed change set and the corresponding test results look good.

Thank you for providing this enhancement to the functionality of the enterprise_sonic resource module collection.

Approved.
